### ISSUE TITLE
General Feedback from JOSS reviews

### ISSUE THREAD

**Issue by avik-pal** [mrazomej/AutoEncode.jl#14]
1. Start with atleast a `0.1.*` versioning. `0.0.*` means every change is breaking
2. Caret specifiers are not needed
3. CUDA Support:
	1. Add CUDA tests. buildkite might be hard for repos in personal accounts, but since the cuda code in the ext involves custom kernels, I would expect tests at the bare minimum.
4. https://github.com/mrazomej/AutoEncoderToolkit.jl/blob/49653858a1e0bb70eb08bd27a18875c59a7c4491/src/layers.jl#L192. Flux now prefers `@layer`
5. Coverage for the project seems extremely low, even excluding the cuda tests it is roughly 60%.
6. > Community guidelines: Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

  Can a section for this be added?
7. Pattern like https://github.com/mrazomej/AutoEncoderToolkit.jl/blob/49653858a1e0bb70eb08bd27a18875c59a7c4491/src/decoders.jl#L452-L456 makes your entire code type-unstable, which might be okay for large networks, but best to avoid that pattern.

**Comment by mrazomej** [comment]
@avik-pal, thank you so much for taking the time to carefully review the paper and the software. I agree with all comments and suggestions and will promptly implement the necessary changes.

I would like to follow up with some questions and clarifications:

> 2. Caret specifiers are not needed

How would you rewrite this example from the `Project.toml` file without the caret specifiers?
```
[compat]
CUDA = "^5.3.3"
ChainRulesCore = "^1.23.0"
```

> 3. CUDA Support:
> Add CUDA tests. buildkite might be hard for repos in personal accounts, but since the cuda code in the ext involves custom kernels, I would expect tests at the bare minimum.

Would it suffice to have a test file that runs locally on my `CUDA`-available machine?

> 7. Pattern like
> 
> [AutoEncoderToolkit.jl/src/decoders.jl](https://github.com/mrazomej/AutoEncoderToolkit.jl/blob/49653858a1e0bb70eb08bd27a18875c59a7c4491/src/decoders.jl#L452-L456)
> Lines 452 to 456 in [4965385](https://github.com/mrazomej/AutoEncoderToolkit.jl/commit/49653858a1e0bb70eb08bd27a18875c59a7c4491)
```julia
 struct JointGaussianLogDecoder <: AbstractGaussianLogDecoder 
     decoder::Flux.Chain 
     µ::Flux.Dense 
     logσ::Flux.Dense 
 end 
```
> makes your entire code type-unstable, which might be okay for large networks, but best to avoid that pattern.

Would rewriting the `struct` as 
```julia
struct JointGaussianLogDecoder{D<:Flux.Chain,L<:Flux.Dense} <: AbstractGaussianLogDecoder
    decoder::D
    µ::L
    logσ::L
end
```
fix the type instability?

Thanks again for your help!

**Comment by avik-pal** [comment]
```julia
[compat]
CUDA = "5.3.3"
ChainRulesCore = "1.23.0"
```

is same as specifying `^...`.

> Would it suffice to have a test file that runs locally on my CUDA-available machine?

Yes, you can commit that. Add a explicit check for `CUDA.functional()` and run those test files if  cuda is functional (which would be false for github actions)

> fix the type instability?

Yes. You can also use https://github.com/jonniedie/ConcreteStructs.jl which makes it easier to write these things.

**Comment by fabian-s** [comment]
Thank you both!

**Comment by mrazomej** [comment]
@avik-pal and @fabian-s 

I implemented the requested changes and released the [`v0.1.0`](https://github.com/mrazomej/AutoEncoderToolkit.jl/releases/tag/v0.1.0) version. The coverage reported does not reflect the package's real coverage. I accidentally kept a lot of "experimental" files in the main branch, but those have been removed. Basically, every function in the available models has been fully tested.

Thanks again, @avik-pal, for the incredibly insightful comments. Your suggestions truly improve the quality of the package.

Please let me know if there is anything else you would like me to implement. If not, we can close this issue.

**Comment by avik-pal** [comment]
The CI isn't running correctly right? It proceeds on error https://github.com/mrazomej/AutoEncoderToolkit.jl/blob/8b80d925ef9a50f928fe746b3195e914730e91ca/.github/workflows/CI.yml#L50 https://github.com/mrazomej/AutoEncoderToolkit.jl/actions/runs/9846125121/job/27183232171#step:7:281

**Comment by mrazomej** [comment]
> The CI isn't running correctly right? It proceeds on error
> 
> https://github.com/mrazomej/AutoEncoderToolkit.jl/blob/8b80d925ef9a50f928fe746b3195e914730e91ca/.github/workflows/CI.yml#L50
> 
> https://github.com/mrazomej/AutoEncoderToolkit.jl/actions/runs/9846125121/job/27183232171#step:7:281

@avik-pal , you are absolutely right, and I apologize for that. I made an error in determining whether or not to test the `CUDA`- related functions on GitHub, but now that is fixed. I changed that `continue-on-error` to `false`.

However, I'm running into an issue I don't really understand. Locally, all tests pass whether `CUDA` is available or not.
```julia
Test Summary:         | Pass  Total     Time
AutoEncoderToolkit.jl | 1387   1387  1m44.3s
```
But on GitHub, there seems to be a recurrent issue when taking gradients with `Zygote` (see example below from [here](https://github.com/mrazomej/AutoEncoderToolkit.jl/actions/runs/9897780064/job/27343079815)). This causes all functions where a gradient is taken to fail. 
```julia
Test Summary:                               | Pass  Error  Total     Time
AutoEncoderToolkit.jl                       | 1344     30   1374  2m54.0s
```
That was the original reason I had set `continue-on-error: true`. I am not sure how to fix this issue.

```julia
NeuralGeodesic training: Error During Test at /Users/runner/work/AutoEncoderToolkit.jl/AutoEncoderToolkit.jl/test/diffgeo.jl:300
  Got exception outside of a @test
  BoundsError: attempt to access 16-element Vector{UInt8} at index [0]
  Stacktrace:
    [1] getindex
      @ ./essentials.jl:13 [inlined]
    [2] isslotfilled
      @ ./dict.jl:162 [inlined]
    [3] skip_deleted
      @ ./dict.jl:706 [inlined]
    [4] iterate
      @ ./dict.jl:724 [inlined]
    [5] _pullback
      @ ~/.julia/packages/Zygote/nsBv0/src/lib/base.jl:53 [inlined]
    [6] _pullback
      @ ~/.julia/packages/Zygote/nsBv0/src/lib/base.jl:70 [inlined]
    [7] merge
      @ ./namedtuple.jl:361 [inlined]
    [8] _pullback(::Zygote.Context{false}, ::typeof(merge), ::@NamedTuple{}, ::Dict{Any, Any})
      @ Zygote ~/.julia/packages/Zygote/nsBv0/src/compiler/interface2.jl:0
    [9] #11
      @ ~/work/AutoEncoderToolkit.jl/AutoEncoderToolkit.jl/src/diffgeo/neural_geodesics.jl:562 [inlined]
   [10] _pullback(ctx::Zygote.Context{false}, f::AutoEncoderToolkit.diffgeo.NeuralGeodesics.var"#11#12"{typeof(AutoEncoderToolkit.diffgeo.NeuralGeodesics.loss), Dict{Any, Any}, RHVAE{AutoEncoderToolkit.VAEs.VAE{JointGaussianLogEncoder{…}, SimpleGaussianDecoder{…}}}, Vector{Float32}}, args::AutoEncoderToolkit.diffgeo.NeuralGeodesics.NeuralGeodesic)
      @ Zygote ~/.julia/packages/Zygote/nsBv0/src/compiler/interface2.jl:0
   [11] pullback(f::Function, cx::Zygote.Context{false}, args::AutoEncoderToolkit.diffgeo.NeuralGeodesics.NeuralGeodesic)
      @ Zygote ~/.julia/packages/Zygote/nsBv0/src/compiler/interface.jl:90
   [12] pullback
      @ ~/.julia/packages/Zygote/nsBv0/src/compiler/interface.jl:88 [inlined]
   [13] withgradient(f::Function, args::AutoEncoderToolkit.diffgeo.NeuralGeodesics.NeuralGeodesic)
      @ Zygote ~/.julia/packages/Zygote/nsBv0/src/compiler/interface.jl:205
   [14] train!(curve::AutoEncoderToolkit.diffgeo.NeuralGeodesics.NeuralGeodesic, rhvae::RHVAE{AutoEncoderToolkit.VAEs.VAE{JointGaussianLogEncoder{…}, SimpleGaussianDecoder{…}}}, t::Vector{Float32}, opt::@NamedTuple{mlp::@NamedTuple{layers::Tuple{@NamedTuple{weight::Optimisers.Leaf{Optimisers.Adam, Tuple{Matrix{Float32}, Matrix{Float32}, Tuple{Float32, Float32}}}, bias::Optimisers.Leaf{Optimisers.Adam, Tuple{Vector{Float32}, Vector{Float32}, Tuple{Float32, Float32}}}, σ::Tuple{}}, @NamedTuple{weight::Optimisers.Leaf{Optimisers.Adam, Tuple{Matrix{Float32}, Matrix{Float32}, Tuple{Float32, Float32}}}, bias::Optimisers.Leaf{Optimisers.Adam, Tuple{Vector{Float32}, Vector{Float32}, Tuple{Float32, Float32}}}, σ::Tuple{}}, @NamedTuple{weight::Optimisers.Leaf{Optimisers.Adam, Tuple{Matrix{Float32}, Matrix{Float32}, Tuple{Float32, Float32}}}, bias::Optimisers.Leaf{Optimisers.Adam, Tuple{Vector{Float32}, Vector{Float32}, Tuple{Float32, Float32}}}, σ::Tuple{}}}}}; loss_function::typeof(AutoEncoderToolkit.diffgeo.NeuralGeodesics.loss), loss_kwargs::Dict{Any, Any}, verbose::Bool, loss_return::Bool)
      @ AutoEncoderToolkit.diffgeo.NeuralGeodesics ~/work/AutoEncoderToolkit.jl/AutoEncoderToolkit.jl/src/diffgeo/neural_geodesics.jl:561
   [15] macro expansion
      @ ~/work/AutoEncoderToolkit.jl/AutoEncoderToolkit.jl/test/diffgeo.jl:324 [inlined]
   [16] macro expansion
      @ ~/hostedtoolcache/julia/1.10.4/x64/share/julia/stdlib/v1.10/Test/src/Test.jl:1577 [inlined]
   [17] top-level scope
      @ ~/work/AutoEncoderToolkit.jl/AutoEncoderToolkit.jl/test/diffgeo.jl:302
   [18] include(fname::String)
      @ Base.MainInclude ./client.jl:489
   [19] macro expansion
      @ ~/work/AutoEncoderToolkit.jl/AutoEncoderToolkit.jl/test/runtests.jl:45 [inlined]
   [20] macro expansion
      @ ~/hostedtoolcache/julia/1.10.4/x64/share/julia/stdlib/v1.10/Test/src/Test.jl:1577 [inlined]
   [21] top-level scope
      @ ~/work/AutoEncoderToolkit.jl/AutoEncoderToolkit.jl/test/runtests.jl:18
   [22] include(fname::String)
      @ Base.MainInclude ./client.jl:489
   [23] top-level scope
      @ none:6
   [24] eval
      @ ./boot.jl:385 [inlined]
   [25] exec_options(opts::Base.JLOptions)
      @ Base ./client.jl:291
   [26] _start()
      @ Base ./client.jl:552
  ```

**Comment by avik-pal** [comment]
Not sure why there is a discrepancy from local runs but doing a merge of `Dict` and `NamedTuple` doesn't seem to have a rrule. Regardless you should do a merge of 2 NamedTuples, `Dict`s generally have unnecessary overheads.

**Comment by mrazomej** [comment]
@avik-pal , that was it! Replacing all `Dict` with `NamedTuple` solved the issue. Thank you so much for your help; now all tests pass on `GitHub.`
```julia
All tests passed!

Test Summary:         | Pass  Total     Time
AutoEncoderToolkit.jl | 1387   1387  8m08.4s
     Testing AutoEncoderToolkit tests passed 
```
I released and registered `v0.1.1` with all these changes. Thanks again for your help!
Please let me know if you would like me to implement anything else.

**Comment by avik-pal** [comment]
Things look great now.
